scripts/fix/diplomats.lua

```diff
@@ -0,0 +1,103 @@
+        pos.id = ent.positions.next_position_id + 1
+        pos.flags.set(entity_position_flags.DO_NOT_CULL)
+        pos.flags.set(entity_position_flags.MENIAL_WORK_EXEMPTION)
+        pos.flags.set(entity_position_flags.SLEEP_PRETENSION)
+        pos.flags.set(entity_position_flags.PUNISHMENT_EXEMPTION)
+        pos.flags.set(entity_position_flags.ACCOUNT_EXEMPT)
+        pos.flags.set(entity_position_flags.DUTY_BOUND)
+        pos.flags.set(entity_position_flags.COLOR)
+        pos.flags.set(entity_position_flags.HAS_RESPONSIBILITIES)
+        pos.flags.set(entity_position_flags.IS_DIPLOMAT)
+        pos.flags.set(entity_position_flags.IS_LEADER)
+        -- not sure what these flags do, but the game sets them for a valid diplomat
+        pos.flags.set(entity_position_flags.unk_12)
+        pos.flags.set(entity_position_flags.unk_1a)
+        pos.flags.set(entity_position_flags.unk_1b)
+        pos.name[0] = "Diplomat"
+        pos.name[1] = "Diplomats"
+        pos.precedence = 70
+        pos.color[0] = 7
+        pos.color[1] = 0
+        pos.color[2] = 1
+    end
+    -- assign responsibilities
+    pos.responsibilities.MAKE_INTRODUCTIONS = true
+    pos.responsibilities.MAKE_PEACE_AGREEMENTS = true
+    pos.responsibilities.MAKE_TOPIC_AGREEMENTS = true
+
+    return pos
+end
+
+
+
+checked = 0
+fixed = 0
+
+for _,ent in pairs(df.world.entities.all) do -- not a thing in lua???
+    if ent.type == 0 and ent.entity_raw.flags.TREE_CAP_DIPLOMACY then
+        checked = checked + 1
+
+        update = true;
+        pos = nil
+        -- see if we need to add a new position or modify an existing one
+        for pos in ent.positions.own do
+            if  pos.responsibilities.MAKE_INTRODUCTIONS and
+                pos.responsibilities.MAKE_PEACE_AGREEMENTS and
+                pos.responsibilities.MAKE_TOPIC_AGREEMENTS then
+                    -- a diplomat position exists with the proper responsibilities - skip to the end
+                    update = false
+                    break
+            end
+            -- Diplomat position already exists, but has the wrong options - modify it instead of creating a new one
+            if pos.code == "DIPLOMAT" then break end
+            pos = nil
+        end
+        if update then
+            -- either there's no diplomat, or there is one and it's got the wrong responsibilities
+            pos = update_pos(pos, ent)
+        end
+
+        -- make sure the diplomat position, whether we created it or not, is set up for proper assignment
+        assign = true
+        for p in ent.positions.assignments do
+            if p.position_id == pos.id then -- it is - nothing more to do here
+                assign = false
+                break
+            end
+        end
+        if assign then -- it isn't - set it up
+            asn = df.entity_position_assignment:new()
+            ent.positions.assignments.insert(0, asn);
+
+            asn.id = ent.positions.next_assignment_id + 1
```

Comment:
You also need to actually change ent.positions.next_assignment_id, like the C++ code. This would be a correct translation:
```lua
asn.id = ent.positions.next_assignment_id
ent.positions.next_assignment_id = ent.positions.next_assignment_id + 1
```